### PR TITLE
Add ondemand table state

### LINKED ISSUE #3210: OnDemand Tables: Create new table state called ONDEMAND
Currently tables are either online or offline. Create a new table state called on-demand where tables are treated as offline unless they are specifically brought online by request. The thought here is that a client that wants to perform a scan or live ingest into a range of KeyExtents will look up the location for these tablets in the TabletLocator. When they are not found, the client contacts the Manager and requests that they be hosted. The client would then wait until the TabletLocator resolves the location for these tablets, then continue.

### PR DESCRIPTION
**Pull request by dlmarion**
This commit adds a new table state called ondemand, which is currently treated like offline. Follow-on commits will need to request that ondemand tablets be hosted for live ingest / immediate scans, modify the tablet servers to unhost ondemand tablets, implement bulk-import into ondemand tables, and more.

Closes #3210

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/security/AuditedSecurityOperation.java [hunk 1]
```diff
@@ -685,8 +685,8 @@ public boolean hasSystemPermission(TCredentials credentials, String user,
       "action: %s; targetTable: %s:%s";
 
   @Override
-  public boolean canOnlineOfflineTable(TCredentials credentials, TableId tableId, FateOperation op,
-      NamespaceId namespaceId) throws ThriftSecurityException {
+  public boolean canOnlineOfflineOnDemandTable(TCredentials credentials, TableId tableId,
```
Comment (EdColeman) [inline server/base/src/main/java/org/apache/accumulo/server/security/AuditedSecurityOperation.java:688]: Could these methods be combined into something like `canChangeTableState(...)` for the method name?  The permission could be a general allowed to change state and not for a specific state.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 2]
```diff
@@ -1086,4 +1086,40 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
     throw new UnsupportedOperationException();
   }
 
+  /**
```
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1089]: Is there a good place to summarize what ondemands means vs on or offline?  I assume the web site would have a more lengthy description, but having something in the javadocs might - would not want to see it on every ondemand method,  but somewhere? (and can be done later)

**Comment by dlmarion** [comment]
`verify` stage is failing due to Maven 3.9.0 issue mentioned in Slack. It passes locally.

**Review by EdColeman — COMMENTED** [review]
I wonder, if, in general if we should manage the state in a more general fashion rather than having specific methods for each state - something along the lines of setState(target) rather than things like  `tableOperations().onDemand()`  We could keep deprecate (but keep) the current setOnline()... by redirecting to the new method with appropriate parameters, but addition states would not need to keep adding specific methods.

**Comment by dlmarion** [comment]
I don't know how many more states we could add. We have online and offline, ondemand is a hybrid of those two. I see what you are saying though. That is something that could be done in a 3.x version before we add this.

File: server/base/src/main/java/org/apache/accumulo/server/security/AuditedSecurityOperation.java (hunk 1 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/security/AuditedSecurityOperation.java]: Renamed to `canChangeTableState`

**Comment by dlmarion** [comment]
There are now 3 PRs up for the OnDemand tablet feature, and I think it's functional although I'm sure there are more edge cases to test. 

  This PR introduces new API methods to put a table in an onDemand state (vs online or offline). The implementation of this creates a Fate operation that ends up running the `ChangeTableState` Fate op. The Manager has been updated such that tablets for a table in the onDemand state will not be hosted, just like an offline table.

  [37](https://github.com/dlmarion/accumulo/pull/37) builds on this and contains modifications for onDemand tablets to be brought online. This is done via the `TabletLocator`, when the client is unable to find a Tablet location and it's an onDemand table, then it makes a call to a new Thrift API method called `bringOnDemandTabletOnline`. When this method is called, the implementation (`TabletClientHandler.bringOnDemandTabletsOnline`) inserts an "ondemand" column into the metadata for the tablet. The `TabletStateChangeIterator` has been modified to return the `TabletLocationState` via the `MetaDataTableScanner` so that the `TabletGroupWatcher` running in the Manager will see the tablet as unassigned and will assign it using the normal tablet assignment process.

  [38](https://github.com/dlmarion/accumulo/pull/38) builds on 37 and contains modifications for onDemand tablets to be unloaded. This is done by a new Thread in the TabletServer that calls the new method `evaluateOnDemandTabletsForUnload` at some interval (defined by a new property). `evaluateOnDemandTabletsForUnload` uses an implementation of a new SPI `OnDemandTabletUnloader` class (default implementation provided) to determine which online tablets for an onDemand tablet should be unloaded. For each tablet returned, the method removes the "ondemand" column from the metadata table for the tablet. Like in 37, the `TabletStateChangeIterator` returns the information to the `TabletGroupWatcher` running in the Manager and it will unload the tablet using the normal process.

**Comment by dlmarion** [comment]
If you want to test this yourself, you can:

  1. Build Accumulo from PR [38](https://github.com/dlmarion/accumulo/pull/38)
  2. Set the following properties in accumulo.properties:
  ```
    manager.tablet.watcher.interval=15s
    table.ondemand.tablet.unloader.interval=1m
    table.custom.ondemand.unloader.inactivity.threshold=120000
  ```
  3. Start up local instance
  4. Log into Accumulo shell, and do the following
  ```
  createtable test
  ```
  > **Note**
  > notice test and 1 tablet in Monitor
  > notice "loc" column in tablet metadata
  ```
  ondemand test
  ```
  > **Note**
  > notice 0 tablets in monitor
  > notice "loc" missing from tablet metadata
  ```
  insert a b c d
  ```
  > **Note**
  > notice shell command wait for tablet to be hosted
  > notice 1 tablet in Monitor
  > notice "loc" column in tablet metadata
  > notice "ondemand" column in tablet metadata
  ```
  addsplits m - test
  ```
  > **Note**
  > notice 1 tablet in Monitor
  > notice "loc" in tablet metadata for source tablet only
  > notice "ondemand" column in tablet metadata for source tablet only
 ```
 sleep 240
 ```
  > **Note**
  > table.custom.ondemand.unloader.inactivity.threshold is set to 2m, waiting 4m should unload the tablet
  > notice 0 tablets in monitor
  > notice "loc" column missing in tablet metadata
  > notice "ondemand" column missing in tablet metadata
  ```
  insert a b c d
 ```
  > **Note**
  > notice shell command wait for tablet to be hosted
  > notice 1 tablet in Monitor
  > notice "loc" column in tablet metadata
  > notice "ondemand" column in tablet metadata

  5. Stop Accumulo
  6. Start Accumulo
  > **Note**
  > test tablet had "ondemand" column on shutdown, so it gets re-hosted
  > notice 1 tablet in Monitor
  > notice "loc" column in tablet metadata
  > notice "ondemand" column in tablet metadata
  7.  Wait a few minutes for tablet to be unloaded
  > **Note**
  > notice 0 tablets in monitor
  > notice "loc" column missing in tablet metadata
  > notice "ondemand" column missing in tablet metadata

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 3]
```diff
@@ -1511,6 +1511,60 @@ public void online(String tableName, boolean wait)
...
+  }
+
+  @Override
+  public void onDemand(String tableName)
+      throws AccumuloSecurityException, AccumuloException, TableNotFoundException {
+    onDemand(tableName, false);
+  }
+
+  @Override
+  public void onDemand(String tableName, boolean wait)
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1547]: Good catch, I wasn't thinking about that when writing this, just trying to create something functional. Updated in 836b227.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 4]
```diff
@@ -1479,36 +1463,90 @@ public void online(String tableName)
...
-      return;
+
+    FateOperation op = null;
+    switch (newState) {
+      case OFFLINE:
+        op = FateOperation.TABLE_OFFLINE;
+        break;
+      case ONDEMAND:
+        op = FateOperation.TABLE_ONDEMAND;
+        if (tableName.equals(MetadataTable.NAME) || tableName.equals(RootTable.NAME)) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1479]: OFFLINE has similar restrictions but does not do these checks client side. For consistency, I think neither or both OFFLINE and ONDEMAND should do the checks client side.

**Review by keith-turner — COMMENTED** [review]
still looking at this, noticed a few things so far

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java (hunk 4 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1479]: Yeah, I was just copying what was there, but I did notice that it was not consistent. Root and Metadata should always be online, right? Enforce client side that OFFLINE and ONDEMAND does *not* work for those tables.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:1482]: Updated in 0751fbf

**Comment by dlmarion** [comment]
Closing this in favor of #3250
